### PR TITLE
Eperez fix verifications

### PR DESCRIPTION
**Pull request by enriquepablo**
Better code for verifications

### CODE REVIEW

File: src/eduid_webapp/email/views.py [hunk 1]
```diff
@@ -184,9 +182,22 @@ def verify(user, code, email):
             '_status': 'error',
             'message': 'emails.code_expired_send_new'
         }
-
+    if code != state.verification.verification_code:
+        msg = "Invalid verification code for: {}".format(state.verification.email)
+        current_app.logger.debug(msg)
```
Comment (fredrikt) [inline src/eduid_webapp/email/views.py]: Could skip the intermediate variable 'msg' here too.

File: src/eduid_webapp/email/views.py [hunk 2]
```diff
@@ -184,9 +182,22 @@ def verify(user, code, email):
...
+        return {
+            '_status': 'error',
+            'message': 'emails.code_invalid'
+        }
     try:
         verify_mail_address(state, proofing_user)
+        current_app.logger.info('Email {} successfully verified for user'
+                                ' {!r}'.format(email, proofing_user))
```
Comment (fredrikt) [inline src/eduid_webapp/email/views.py]: Use {} not {!r} to log user objects.

File: src/eduid_webapp/email/views.py [hunk 3]
```diff
@@ -213,9 +218,13 @@ def verify_link(user):
     email = request.args.get('email')
     if code and email:
         current_app.logger.debug('Trying to save email address {} as verified for user {}'.format(email, proofing_user))
-
         db = current_app.proofing_statedb
-        state = db.get_state_by_eppn_and_email(proofing_user.eppn, email)
+        state = db.get_state_by_eppn_and_email(proofing_user.eppn, email, raise_on_missing=False)
+
+        if state is None:
+            current_app.logger.info("Verification code unknown: {}. Sending new code".format(code))
```
Comment (fredrikt) [inline src/eduid_webapp/email/views.py]: It's not the code that is unknown, there is no state for this eppn+email - right?

File: src/eduid_webapp/email/views.py [hunk 4]
```diff
@@ -227,15 +236,14 @@ def verify_link(user):
         if code != state.verification.verification_code:
             current_app.logger.warning("Invalid verification code for: {}".format(state.verification.email))
             return redirect(current_app.config['SAML2_LOGIN_REDIRECT_URL'])
-
         try:
             verify_mail_address(state, proofing_user)
+            current_app.logger.info('Verified email {} for user {!r}'.format(email, user))
+            return redirect(current_app.config['SAML2_LOGIN_REDIRECT_URL'])
```
Comment (fredrikt) [inline src/eduid_webapp/email/views.py]: This return looks redundant with the one after the try: clause if you were to undo the indent-changes of the one you moved into the except: ?

File: src/eduid_webapp/phone/verifications.py [hunk 5]
```diff
@@ -71,3 +74,36 @@ def send_verification_code(user, phone):
...
+    try:
+        proofing_user.phone_numbers.add(new_phone)
+    except DuplicateElementViolation:
+        proofing_user.phone_numbers.find(number).is_verified = True
+        if has_primary is None:
+            proofing_user.phone_numbers.find(number).is_primary = True
+
+    save_and_sync_user(proofing_user)
+    current_app.logger.info('Mobile {!r} confirmed '
+                            'for user {!r}'.format(number, proofing_user))
```
Comment (fredrikt) [inline src/eduid_webapp/phone/verifications.py]: Log user with {}

File: src/eduid_webapp/phone/views.py [hunk 6]
```diff
@@ -177,7 +177,15 @@ def verify(user, code, number):
                              'for user {!r}'.format(number, proofing_user))
 
     db = current_app.proofing_statedb
-    state = db.get_state_by_eppn_and_mobile(proofing_user.eppn, number)
+    state = db.get_state_by_eppn_and_mobile(proofing_user.eppn, number,
+            raise_on_missing=False)
+    if state is None:
+        current_app.logger.debug('Invalid verification code for phone {!r}'
```
Comment (fredrikt) [inline src/eduid_webapp/phone/views.py]: Rather "Couldn't find proofing state for user {} and number {}"

File: src/eduid_webapp/phone/views.py [hunk 7]
```diff
@@ -187,48 +195,29 @@ def verify(user, code, number):
...
-            proofing_user.phone_numbers.find(number).is_primary = True
-
     try:
-        save_and_sync_user(proofing_user)
+        verify_phone_number(state, proofing_user)
+        current_app.logger.info('phone number {!r} successfully verified'
+                                 ' for user {!r}'.format(number, proofing_user))
+        phones = {
+                'phones': proofing_user.phone_numbers.to_list_of_dicts(),
+                'message': 'phones.verification-success'
```
Comment (fredrikt) [inline src/eduid_webapp/phone/views.py]: All the other messages use underscore, while this one (and the same one for emails) uses hyphens - intentional?

File: src/eduid_webapp/phone/views.py [hunk 8]
```diff
@@ -187,48 +195,29 @@ def verify(user, code, number):
...
-    except DuplicateElementViolation:
-        proofing_user.phone_numbers.find(number).is_verified = True
-        if has_primary is None:
-            proofing_user.phone_numbers.find(number).is_primary = True
-
     try:
-        save_and_sync_user(proofing_user)
+        verify_phone_number(state, proofing_user)
+        current_app.logger.info('phone number {!r} successfully verified'
+                                 ' for user {!r}'.format(number, proofing_user))
```
Comment (fredrikt) [inline src/eduid_webapp/phone/views.py]: Capitalize sentence, and log user with {}.

File: src/eduid_webapp/phone/views.py [hunk 9]
```diff
@@ -187,48 +195,29 @@ def verify(user, code, number):
...
-
     if code != state.verification.verification_code:
         msg = "Invalid verification code: {!r}".format(state.verification)
         current_app.logger.debug(msg)
         return {
             '_status': 'error',
             'message': 'phones.code_invalid'
         }
-
-    current_app.proofing_statedb.remove_state(state)
```
Comment (fredrikt) [inline src/eduid_webapp/phone/views.py]: I don't see phone state removed anywhere now?

File: src/eduid_webapp/email/views.py (hunk 1 above)
Comment (enriquepablo) [reply src/eduid_webapp/email/views.py]: Done

File: src/eduid_webapp/email/views.py (hunk 2 above)
Comment (enriquepablo) [reply src/eduid_webapp/email/views.py]: done

File: src/eduid_webapp/email/views.py (hunk 3 above)
Comment (enriquepablo) [reply src/eduid_webapp/email/views.py]: Fixed

File: src/eduid_webapp/email/views.py (hunk 4 above)
Comment (enriquepablo) [reply src/eduid_webapp/email/views.py]: fixed

File: src/eduid_webapp/phone/verifications.py (hunk 5 above)
Comment (enriquepablo) [reply src/eduid_webapp/phone/verifications.py]: fixed

File: src/eduid_webapp/phone/views.py (hunk 6 above)
Comment (enriquepablo) [reply src/eduid_webapp/phone/views.py]: fixed

File: src/eduid_webapp/phone/views.py (hunk 7 above)
Comment (enriquepablo) [reply src/eduid_webapp/phone/views.py]: It's the string that is registered as msg id in the react app. There is no restriction, other than using a short string. There are some with spaces...

File: src/eduid_webapp/email/views.py [hunk 10]
```diff
@@ -184,9 +182,22 @@ def verify(user, code, email):
             '_status': 'error',
             'message': 'emails.code_expired_send_new'
         }
-
+    if code != state.verification.verification_code:
```
Comment (fredrikt) [inline src/eduid_webapp/email/views.py]: Hmm, when you now check for a matching code _after_ you check for timeout, a new code will be sent out even if the wrong code is supplied. I'm a bit unsure what is possible to do through the UI, but it appears incorrect to check timeout before checking the code.

File: src/eduid_webapp/phone/views.py (hunk 9 above)
Comment (enriquepablo) [reply src/eduid_webapp/phone/views.py]: It is done in the `verify_phone_number` function, in the `eduid_webapp.phones.verifications` module.

File: src/eduid_webapp/phone/views.py (hunk 8 above)
Comment (enriquepablo) [reply src/eduid_webapp/phone/views.py]: fixed

File: src/eduid_webapp/email/views.py [hunk 11]
```diff
@@ -213,29 +216,31 @@ def verify_link(user):
...
     if code and email:
         current_app.logger.debug('Trying to save email address {} as verified for user {}'.format(email, proofing_user))
-
         db = current_app.proofing_statedb
-        state = db.get_state_by_eppn_and_email(proofing_user.eppn, email)
+        state = db.get_state_by_eppn_and_email(proofing_user.eppn, email, raise_on_missing=False)
+
+        if state is None:
+            current_app.logger.info("Missing state for verification code received for email {} "
+                                    "and user {}.  Sending new code.".format(email, user))
```
Comment (fredrikt) [inline src/eduid_webapp/email/views.py]: I think we're not really sending a new code here?

File: src/eduid_webapp/email/views.py [hunk 12]
```diff
@@ -213,29 +216,31 @@ def verify_link(user):
...
+            return redirect(current_app.config['SAML2_LOGIN_REDIRECT_URL'])
 
         timeout = current_app.config.get('EMAIL_VERIFICATION_TIMEOUT', 24)
         if state.is_expired(timeout):
             current_app.logger.info("Verification code is expired for: {}. Sending new code".format(
```
Comment (fredrikt) [inline src/eduid_webapp/email/views.py]: The call to send_verification_code is removed below, so log message would be misleading.

File: src/eduid_webapp/phone/verifications.py [hunk 13]
```diff
@@ -69,5 +72,38 @@ def send_verification_code(user, phone):
...
+        proofing_user.phone_numbers.add(new_phone)
+    except DuplicateElementViolation:
+        proofing_user.phone_numbers.find(number).is_verified = True
+        if has_primary is None:
+            proofing_user.phone_numbers.find(number).is_primary = True
+
+    save_and_sync_user(proofing_user)
+    current_app.logger.info('Mobile {} confirmed '
+                            'for user {}'.format(number, proofing_user))
+    current_app.stats.count(name='mobile_verify_success', value=1)
```
Comment (fredrikt) [inline src/eduid_webapp/phone/verifications.py]: Log message and stats talks about mobile - should be phone now. Seems to apply to a number of other log messages too. Let's keep the stats keys for now though - needs to be renamed in the stats backend at the time of update to keep old stats.

File: src/eduid_webapp/email/views.py (hunk 11 above)
Comment (enriquepablo) [reply src/eduid_webapp/email/views.py]: You are right

Comment (enriquepablo) [reply src/eduid_webapp/email/views.py]: fixed

File: src/eduid_webapp/email/views.py (hunk 12 above)
Comment (enriquepablo) [reply src/eduid_webapp/email/views.py]: fixed

File: src/eduid_webapp/phone/verifications.py (hunk 13 above)
Comment (enriquepablo) [reply src/eduid_webapp/phone/verifications.py]: I've fixed all log messages

**Pull request merged**